### PR TITLE
Crafting system API added

### PR DESCRIPTION
**Pull request by guilhermehto**
UI is still missing. If you run this demo now you'll see some logs on the console and that's it. It's possible to play around with the API already by creating new items and recipes. 

I'm PRing to get feedback on the API: `CraftingStation.gd`. 

It works kind of like a merchant, but instead of coins, you're giving ingredients :)

If you guys could give me some feedback it would be great @razcore-art @NathanLovato 

ps: Grammarly just alerted me that the correct is items and not _itens_, I'll fix the spelling mistakes on a future commit.

### CODE REVIEW

File: 01-17-crafting-system/crafting/CraftingStation.gd [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+
+export var craftable_items = []
+
+
+func can_craft_item(itens : Array, recipe : Array) -> int:
+	"""
+	Returns the amount of a certain item that can be crafted given a recipe and an array of itens
+	"""
+	var amount : = 999999
```
Comment (NathanLovato) [inline 01-17-crafting-system/crafting/CraftingStation.gd]: Set the amount to `0` and remove `can_craft`. If the amount is 0, you can't craft anything. It's weird to have this function return a negative amount.

File: 01-17-crafting-system/crafting/CraftingStation.gd [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		return -1
+
+
+func get_craftables(itens : Array) -> Array:
+	"""
+	Returns an array of the possible craftable items with the given itens
+	"""
+	var craftables : = []
+	for craftable_item in craftable_items:
+		craftable_item = craftable_item as Item
```
Comment (NathanLovato) [inline 01-17-crafting-system/crafting/CraftingStation.gd]: I wouldn't add that line for just one statement that uses it: `craftable_item.recipe`. At runtime, the items should already have the corresponding type so you're adding an extra operation.

File: 01-17-crafting-system/crafting/items/Item.gd [hunk 3]
```diff
@@ -0,0 +1,6 @@
+extends Resource
+class_name Item
+
+export var item_name : String = ""
+export var recipe : Array = []
```
Comment (NathanLovato) [inline 01-17-crafting-system/crafting/items/Item.gd]: Needs some docs. This is an array of ingredients?

File: 01-17-crafting-system/crafting/items/Item.gd [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+export var crafted_item : PackedScene
```
Comment (NathanLovato) [inline 01-17-crafting-system/crafting/items/Item.gd]: I think going back to JSON will make things more explicit. Be sure to name the key/property like something more explicit: `crafted_item` is a scene, it's not a crafted item. So better call it something like `Item.scene`

**Review by NathanLovato — CHANGES_REQUESTED** [review]
You can use json if you want moving forward. Or CSV tables, it's also nice for things like item databases. Because Resources have some issues.

**Comment by razcore-rad** [comment]
@NathanLovato I'd say that csv is a bit of a weird place in godot cause if the engine finds a csv file it tries to load it as a translation and will fails so all sorts of whistles and bells start ringing in the bottom docker.

JSON is really nice to work with because it's easy to convert `Dictionary` and `Array` to JSON, that said, `Resource`s are also nice cause we can use them in the inspector and if we stick with a simple rule: once a resource is modified via inspector and happy with it, never touch it again except through code.

Alright, that said, let's get back to business. This is how I'd approach making the crafting station:
```gdscript
# Item.gd
extends Resource
class_name Item


export var name : = ""
export var recipe : = {}
```

```gdscript
# Game.gd
extends Node


onready var crafting_station : CraftingStation = $CraftingStation


var items : = {
	wood = preload("res://crafting/items/Wood.tres"),
	iron_ore = preload("res://crafting/items/IronOre.tres"),
	iron_ingot = preload("res://crafting/items/IronIngot.tres")
}


func _ready() -> void:
	var recipe : = {
		items.wood: 50,
		items.iron_ore: 50,
		items.iron_ingot: 5
	}
	var recipe_new : = crafting_station.craft(items.iron_ingot, recipe, 2)
	for it in recipe_new:
		print(it.name, " ", recipe_new[it])
```

```gdscript
# CraftingStation.gd
extends Node
class_name CraftingStation
"""
Responsible to craft items based on recipes
"""

export var available_items = []


func can_craft_1(item: Item, recipe: Dictionary) -> bool:
	var out : = true
	if item.recipe.empty() or not (item in available_items and recipe.has_all(item.recipe.keys())):
		out = false
	else:
		for recipe_item in recipe:
			if item.recipe.has(recipe_item) and recipe[recipe_item] < item.recipe[recipe_item]:
				out = false
				break
	return out


func can_craft(item: Item, recipe: Dictionary, amount: int = 1) -> bool:
	if amount < 1:
		return false
	elif amount == 1:
		return can_craft_1(item, recipe)
	return can_craft(item, use(item, recipe), amount - 1)


# typed GDScript doesn't like Resource keys in Dictionary
# https://github.com/godotengine/godot/issues/25088
# func craft_1(item: Item, recipe: Dictionary) -> Dictionary:
func craft_1(item, recipe: Dictionary) -> Dictionary:
	var out : = recipe.duplicate()
	if can_craft(item, out):
		if not out.has(item):
			out[item] = 0
		out[item] += 1
		out = use(item, out)
	return out


func craft(item: Item, recipe: Dictionary, amount: int = 1) -> Dictionary:
	if not (amount >= 1 and can_craft(item, recipe, amount)):
		return recipe
	return craft(item, craft_1(item, recipe), amount - 1)


func use(item: Item, recipe: Dictionary) -> Dictionary:
	var out : = recipe.duplicate()
	if out.has_all(item.recipe.keys()):
		for it in item.recipe:
			out[it] -= item.recipe[it]
	return out
```

With the caveat that recursion isn't necessarily the best solution. If you want to make more than 1024 items it will fail cause of stack overflow. I just like recursion it cause it's elegant. If you don't understand the code let me know, I can run you through it.

But basically `recipe` is a `Dictionary` in my case, the keys are `Item`s and values are amount of `Item`s. Whenever we try crafting a new item, we create a new updated `Dictionary` if the `CraftingStation` can craft it, if not we return the same `Dictionary` back.

**edit**: updated to remove `can_be_crafted` from `Item.gd` and perhaps make the `if` expressions clear - don't know for sure.

**Comment by razcore-rad** [comment]
In case you're wondering, yes, this approach is much slower than mutating the `Dictionary` etc. But this is **good** practice and once we're happy with how the implementation works, then we can optimize. Without doing optimization from beginning we have a better understanding of how to solve the problem.

If this is as example then it should be mentioned somewhere.

**Comment by razcore-rad** [comment]
Here, this would be already a huge boost (in languages like GDScript recursion is really not a nice option - performance-wise):

```gdscript
# CraftingStation.gd
extends Node
class_name CraftingStation
"""
Responsible to craft items based on recipes
"""

export var items = []


func can_craft(item: Item, recipe: Dictionary, amount: int = 1) -> bool:
	var out : = true
	if amount < 1 or item.recipe.empty() or not (item in items and recipe.has_all(item.recipe.keys())):
		out = false
	else:
		var item_adjusted : = adjust_item_recipe(item, amount)
		for it in item.recipe:
			if recipe[it] < item_adjusted.recipe[it]:
				out = false
				break
	return out

## typed GDScript doesn't like Resource keys in Dictionary
## https://github.com/godotengine/godot/issues/25088
# func craft(item: Item, recipe: Dictionary, amount: int = 1) -> Dictionary:
func craft(item, recipe: Dictionary, amount: int = 1) -> Dictionary:
	var out : = recipe.duplicate()
	if amount > 0 and can_craft(item, recipe, amount):
		out[item] = out.get(item, 0) + amount
		out = use(adjust_item_recipe(item, amount), out)
	return out


func adjust_item_recipe(item: Item, amount: int) -> Item:
	var out : Item = item.duplicate()
	if amount > 1:
		for it in out.recipe:
			out.recipe[it] *= amount
	return out


func use(item: Item, recipe: Dictionary) -> Dictionary:
	var out : = recipe.duplicate()
	if out.has_all(item.recipe.keys()):
		for it in item.recipe:
			if out[it] < item.recipe[it]:
				out = recipe
				break
			out[it] -= item.recipe[it]
	return out
```

We make use of the fact that making `x` amount of `Item`s, is exactly as making **one** `Item` with its recipe adjusted by `x`. This is a much nicer solution, I guess, earlier I just tried recursion-based cause I wanted to see how it works with GDScript :)

**Comment by razcore-rad** [comment]
Oh yah and as a general rule, if you have a function name **`can_`something**, **`is_`something**, etc. then be sure to return a `bool`, otherwise it can be confusing. There are exceptions sure, but generally speaking, can/is implies true/false so we should keep to them.

**Comment by NathanLovato** [comment]
Your feedback is looking good. @guilhermehto feel free to reuse Razvan's code, with clear names (no `var it` in tutorials on the channel) if you want

**Comment by guilhermehto** [comment]
Hey, thanks for all the inputs! :)

I'm currently converting the code to use json, then I'll port what @razcore-art proposed.

**Comment by guilhermehto** [comment]
I made the system use json now. It's waaay easier to create new items and it doesn't feel like a chore anymore. 

I used most of @razcore-art's implementation but made some tweaks to it. 
If any of you have feedback feel free to let me know, if not, I think this is okay to merge. I decided to let UI out of it. It's going to be complex, and as you said, there are enough UI tutorials on the channel as of now. Unless you want me to cover it :) 

@NathanLovato pinging you so you can see the changes

**Pull request merged**